### ISSUE TITLE
Extra image crop focus: face

### ISSUE THREAD

**Issue by crystalenka** [akeeba/social-magick#10]
There are some instances where it would be beneficial to crop an image based on the location of a face, instead of from the center. (Author profile pages, for example)

The easiest way is probably going to be to allow someone to sign up for a 3rd-party service like Rapid API (https://rapidapi.com/inferdo/api/face-detection6) and only allow it as an option if there's a valid API key.

The linked Face Detection API is free up to 5,000 requests a month - more than enough for most sites.

They offer sample code for cURL requests.

**Comment by nikosdion** [comment]
## Multiple backend

We can have different backends to the face detection code:
* [ ] [RapidAPI Face Detection](https://rapidapi.com/inferdo/api/face-detection6?endpoint=apiendpoint_430e347f-7eca-4190-9e1b-e6e6114879e0)
* [ ] [Microsoft Azure Face](https://docs.microsoft.com/en-us/rest/api/faceapi/face/detectwithstream)
* [ ] [AWS Rekognition](https://docs.aws.amazon.com/rekognition/latest/dg/API_DetectFaces.html)
* [ ] [PHP Facedetect](https://github.com/infusion/PHP-Facedetect) — Really only applies to dedicated servers. I've not seen a host which provides this extension by default.

Having multiple backends is a bit more flexible, especially if the client is already heavily invested in AWS or Azure. RapidAPI is a good solution for sites which don't already use a public cloud. PHP Facedetect is great for advanced and privacy-conscious users who would rather compile and configure a PHP extension than use a public cloud.

## Further notes

* All backends can provide information about one or more faces. We will keep the largest / most confident one — depending on which of these two pieces of information the backend provides.
* All backends can provide a bounding box for the detected face (x, y, width, height). This is the data I will be using.
* We need to cache the results keyed by image location and a unique identifier which comes from the image file length and last modified date (so when the image changes we can query its properties again).
* When cropping we will expand/shrink the face's bounding box (from its centre) to match the requested image crop dimensions width and height. This will have the effect on focusing on the face.

**Comment by SemaphoreOxalis** [comment]
A focus lib seems more reliable like : https://www.npmjs.com/package/image-focus

**Comment by SemaphoreOxalis** [comment]
Obviously, I was talking about "something like that" (vs visage detection) not that specific node/npm etc I knwo that PHP Kirby CMS offert this function. Maybe we can look there how they achive that ;-)

**Comment by nikosdion** [comment]
If you look at the documentation in https://github.com/felixkoch/PHP-FaceDetector you will see that it **DOES** detect faces and **DOES** return the coordinates and size of the detected face. Why make life harder?

**Comment by SemaphoreOxalis** [comment]
Because all images are not about face ?

**Comment by nikosdion** [comment]
Um, you may want to read the issue's title again. This issue is **SPECIFICALLY** about adding a mode which allows this plugin to focus **ON A FACE**. There's a very clear use case for this. Business sites, blogs, etc have dedicated pages with a bio of key members / editors. Usually, these pages have the face of that person, and most of the times it is in portrait mode. It makes perfect sense to use something like the Overlay template to create an OpenGraph image preview focused on the face of that person. That's what this issue is really all about.

Finding the focus point of any random image is a different problem. I wouldn't solve it in pure PHP. I'd use an LLM to analyse the image and give me the approximate percentage of the width and height of the image where the visual focus is. Solving that problem does require this kind of solution. But detecting a face? It's a much simpler problem. One which can be solved without spending five cents a pop – which might very soon become fifty or a hundred, when the AI bubble bursts and companies start charging what it costs them to run inference instead of a subsidised price.

OK? Different problems, different solutions, different process to solve them. I am currently NOT interested in the generic "where's the focus point of this image?" problem. It's hardly a problem that's worth addressing in the context of this plugin because, you know, you can just use a different template for fairly expansive images.

**Comment by SemaphoreOxalis** [comment]
Got it, thanks for reminding me that this issue is about automatic face detection. My earlier suggestion was a bit broader and somewhat tangential: I was thinking about allowing manual selection of the focus point on the chosen image — whether that’s a face, a crowd, or a monument.

I completely agree about not introducing any dependency on an LLM, Just sharing the idea in case it’s useful — happy to stay focused on the face-detection scope if that’s the priority.

**Comment by nikosdion** [comment]
I understand what you have in mind. It is something that's already supported, in a sense. You can already work it backwards with the nudge parameters. Can this instead be written as a focus point so we can work out the nudge parameters automatically? Sure. Can you select the focus point interactively? Not really, because that would require support in Joomla's media manager.

The biggest problem is that this plugin has not been documented very well, making it really hard to understand what you can do with it. Be patient. I took it over after it was languishing in near-death for three years. I have ideas, but they will take time to implement.
